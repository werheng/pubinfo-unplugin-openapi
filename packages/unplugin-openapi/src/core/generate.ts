import { join } from 'node:path'
import { cwd } from 'node:process'
import { generateService as gen } from '@umijs/openapi'
import type { Options } from '../types'

export function generateOpenAPI(options: Pick<Options, 'imports' | 'input' | 'output' >, root = cwd()) {
  const { imports, input, output } = options

  const outputStrs = (output as string).split('/')
  const projectName = outputStrs.pop()
  const serversPath = outputStrs.length > 0 ? outputStrs.join('/') : ''

  gen({
    requestLibPath: imports,
    templatesFolder: join(root, 'node_modules/@pubinfo/unplugin-openapi/templates'),
    serversPath,
    projectName,
    schemaPath: input?.startsWith('http') ? input : join(root, input as string),
  })
}
