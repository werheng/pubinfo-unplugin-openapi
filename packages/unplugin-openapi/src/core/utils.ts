import { createHash } from 'node:crypto'
import type { Buffer } from 'node:buffer'
import { dirname } from 'node:path'
import { promises as fs } from 'node:fs'
import type { Arrayable } from '../types'

export function slash(str: string) {
  return str.replace(/\\/g, '/')
}

export function arraify<T>(target: Arrayable<T>) {
  return Array.isArray(target) ? target : [target]
}

export function getHash(text: Buffer | string, length = 8): string {
  const h = createHash('sha256').update(text).digest('hex').substring(0, length)
  if (length <= 64)
    return h
  return h.padEnd(length, '_')
}

export async function writeFile(filePath: string, content = '') {
  await fs.mkdir(dirname(filePath), { recursive: true })
  return await fs.writeFile(filePath, content, 'utf-8')
}
