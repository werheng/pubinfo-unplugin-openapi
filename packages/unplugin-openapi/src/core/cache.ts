import { existsSync, unlinkSync } from 'node:fs'
import { join } from 'node:path'
import fg from 'fast-glob'
import type { CacheOptions } from '../types'
import { getHash, writeFile } from './utils'

export default function createCache(options: CacheOptions, root: string) {
  const { cacheDir } = options

  function genCacheKey(id: string, content: string) {
    return `${getHash(id)}_${getHash(content)}`
  }

  function getIdFromKey(key: string) {
    return key.split('_')[0]
  }

  function hasCache(key: string) {
    return existsSync(join(cacheDir, `${key}.json`))
  }

  function setCache(key: string, content: string) {
    removeCache(key)
    writeFile(join(cacheDir, `${key}.json`), content)
  }

  async function removeCache(key: string) {
    const id = getIdFromKey(key)
    const result = await fg(
      join(cacheDir, `${id}_**.json`),
      {
        absolute: true,
        cwd: root,
        onlyFiles: true,
        followSymbolicLinks: true,
      },
    )

    if (result.length > 0)
      result.forEach(unlinkSync)
  }

  return {
    genCacheKey,
    hasCache,
    setCache,
    removeCache,
  }
}
