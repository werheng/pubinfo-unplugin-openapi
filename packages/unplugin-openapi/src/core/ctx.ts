import { cwd } from 'node:process'
import { existsSync, readFileSync } from 'node:fs'
import { join } from 'node:path'
import http from 'node:http'
import https from 'node:https'
import { log } from 'node:console'
import fetch from 'node-fetch'
import type { Options } from '../types'
import { generateOpenAPI } from './generate'
import { slash } from './utils'
import createCache from './cache'

export function createContext(rawOptions: Options, root = cwd()) {
  root = slash(root)

  const options = {
    imports: 'import request from \'../index\'',
    output: './src/api/service',
    watch: true,
    force: false,
    enabled: true,
    batch: [],
    ...rawOptions,
  }

  const batch = options.batch.length > 0 ? options.batch : [options]
  const dirs = batch.map(opt => (opt.input || options.input || '').replace(/^\.\/|\.\.\//g, '**/'))

  const { hasCache, setCache, genCacheKey } = createCache({
    root,
    cacheDir: join(root, 'node_modules/.pubinfo-openapi'),
  })

  async function generateTS() {
    await Promise.all(batch.map(async (opt) => {
      const mergeOptions = { ...options, ...opt }
      const outputPath = join(root, mergeOptions.output)

      if (!mergeOptions.input || !mergeOptions.enabled)
        return

      const openAPI = await getSchema(mergeOptions.input)
      if (!openAPI) {
        log('openapi config is empty')
        return
      }

      const cacheKey = genCacheKey(mergeOptions.input, openAPI)

      if (hasCache(cacheKey) && existsSync(outputPath) && !mergeOptions.force)
        return

      await generateOpenAPI(mergeOptions, root)
      setCache(cacheKey, openAPI)
    }))
  }

  return {
    root,
    dirs,
    generateTS,
  }
}

async function getSchema(schemaPath: string) {
  if (schemaPath.startsWith('http')) {
    const protocol = schemaPath.startsWith('https:') ? https : http
    try {
      const agent = new protocol.Agent({ rejectUnauthorized: false })
      const data = await fetch(schemaPath, { agent }).then(res => res.text())
      return data
    }
    catch (error) {
      log('fetch openapi error:', error)
    }
    return null
  }

  const schema = readFileSync(schemaPath, { encoding: 'utf-8' })
  return schema
}
