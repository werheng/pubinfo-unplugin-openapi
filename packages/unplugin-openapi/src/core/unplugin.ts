import { createUnplugin } from 'unplugin'
import { minimatch } from 'minimatch'
import type { Options } from '../types'
import { createContext } from './ctx'
import { slash } from './utils'

export default createUnplugin<Options>((options) => {
  let config: any
  return {
    name: 'unplugin-openapi',
    enforce: 'post',
    async buildStart() {
      if (config.command === 'serve') {
        const { generateTS } = await createContext(options)
        await generateTS()
      }
    },
    vite: {
      configResolved(resolvedConfig) {
        config = resolvedConfig
      },
      async handleHotUpdate({ file }) {
        const { dirs, generateTS } = await createContext(options)
        if (dirs?.some(glob => minimatch(slash(file), slash(glob))))
          await generateTS()
      },
    },
  }
})
