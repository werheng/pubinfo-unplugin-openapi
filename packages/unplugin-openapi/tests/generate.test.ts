import { join, resolve } from 'node:path'
import { readFile, readdir } from 'node:fs/promises'
import { describe, expect, it } from 'vitest'
import { createContext } from '../src/core/ctx'
import type { Options } from '../src/types'

const root = resolve(__dirname, '../')
async function generateFile(options: Options) {
  const ctx = createContext(options, root)
  await ctx.generateTS()

  const dirs = await readdir(join(root, options.output!))
  return await Promise.all(dirs.map(async (file) => {
    return await readFile(join(root, options.output!, file), { encoding: 'utf-8' })
  }))
}

describe('generate', async () => {
  it('should generate by openapi.json', async () => {
    const results = await generateFile({
      templates: './templates',
      input: './tests/inputs/openapi-example.json',
      output: './tests/outputs/openapi-example',
      force: true,
    })

    results.forEach((result) => {
      expect(result).toMatchSnapshot()
    })
  })

  it('should generate by swagger.json', async () => {
    const results = await generateFile({
      templates: './templates',
      input: './tests/inputs/swagger-example.json',
      output: './tests/outputs/swagger-example',
      force: true,
    })

    results.forEach((result) => {
      expect(result).toMatchSnapshot()
    })
  })
})
