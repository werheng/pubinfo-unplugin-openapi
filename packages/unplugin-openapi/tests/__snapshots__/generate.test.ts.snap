// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > should generate by openapi.json 1`] = `
"import request from '@/api';

/** 分类详情查询 GET /address/category/\${param0} */
export function save(params: API.saveParams, options?: { [key: string]: any }) {
  const { id: param0, ...queryParams } = params;
  return request.Get<API.ResponseData>(\`/address/category/\${param0}\`, {
    params: { ...queryParams },
    ...(options || {}),
  });
}
"
`;

exports[`generate > should generate by openapi.json 2`] = `
"// API 更新时间：
// API 唯一标识：
export * from './biaozhundezhimokuai';
"
`;

exports[`generate > should generate by openapi.json 3`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
declare namespace API {
  
        interface ResponseData
          {
                'code'?: number;
                'msg'?: string;
                'hint'?: string;
                'data'?: Record<string, any>;
                'success'?: boolean;
          }
        
  
        interface saveParams
          {
                'id': number;
          }
        
  
}
"
`;

exports[`generate > should generate by swagger.json 1`] = `
"import request from '@/api';

/** 登录 POST /login */
export function loginUsingPost(body: API.LoginBody, options?: { [key: string]: any }) {
  return request.Post<API.RObject_>('/login', body, {
    headers: {
      'Content-Type': 'application/json',
    },
    ...(options || {}),
  });
}

/** 登出 DELETE /logout */
export function logoutUsingDelete(options?: { [key: string]: any }) {
  return request.Delete<API.RObject_>(
    '/logout',
    {},
    {
      ...(options || {}),
    },
  );
}

/** 刷新 POST /refresh */
export function refreshUsingPost(options?: { [key: string]: any }) {
  return request.Post<API.RObject_>(
    '/refresh',
    {},
    {
      ...(options || {}),
    },
  );
}

/** 注册 POST /register */
export function registerUsingPost(body: API.RegisterBody, options?: { [key: string]: any }) {
  return request.Post<API.RObject_>('/register', body, {
    headers: {
      'Content-Type': 'application/json',
    },
    ...(options || {}),
  });
}
"
`;

exports[`generate > should generate by swagger.json 2`] = `
"// API 更新时间：
// API 唯一标识：
export * from './denglujiekou';
"
`;

exports[`generate > should generate by swagger.json 3`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
declare namespace API {
  
        interface LoginBody
          {
                'password'?: string;
                'username'?: string;
          }
        
  
        interface RegisterBody
          {
                'password'?: string;
                'username'?: string;
          }
        
  
        interface RObject_
          {
                'code'?: number;
                'data'?: Record<string, any>;
                'msg'?: string;
          }
        
  
}
"
`;
